### PR TITLE
Add Friend classes to C# with the help of Analyzers and Attributes.

### PR DESCRIPTION
**Pull request by gradientvera**
So basically, because of ECS every member in components is public and anything can write/invoke them even if they shouldn't, yes? The solution we came up with was to code better comments/xmldocs, and review things properly.

But of course, the dumb coder will still write to/invoke them directly instead of going through the intended entity system... The drunk maintainer will absolutely miss these awful crimes against humanity, and the codebase will burn and suffer as a result. All because there's no encapsulation. Smh.

Because of these awful predictions about the future, I decided to sacrifice myself and learn roslyn analyzers with the goal of avoiding that timeline and save the project from the shitcoders: 
I implemented "friend classes" in C#.

The way this works is very simple. You add a `[Friend(typeof(MyFriend), typeof(MyBestFriend))]` attribute to your component, and now only those two classes will be able to write/invoke members in it. Any class can *read*, but only friends can write/invoke.
This also works with inheritance: If your base class is friends with Foo, and the inheriting class is friends with Bar, the inheriting class is friends with Foo as well, but the base class is not friends with Bar. In short, this should probably force you to think about your relationships between components/things using components..

Shoutouts to Y for giving me the motivation required for coding this.

### CODE REVIEW

**Comment by Ygg01** [comment]
Motivation? I believe it was anguish.

File: Robust.Analyzers/FriendAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                    break;
+
+                // Otherwise, do nothing.
+                default:
+                    return;
+            }
+
+            // Get the friend attribute, if it can't be found do nothing...
+            if (context.Compilation.GetTypeByMetadataName(FriendAttribute) is not {} friendAttr)
+                return;
```
Comment (PaulRitter) [inline Robust.Analyzers/FriendAnalyzer.cs]: this is somewhat sus, i'd at least throw a warning here?

**Review by PaulRitter — APPROVED** [review]
very non-invasive, meaning easily reworkable if we find a better solution down the line.

i like it

File: Robust.Analyzers/Robust.Analyzers.csproj [hunk 2]
```diff
@@ -1,7 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
Comment (PaulRitter) [inline Robust.Analyzers/Robust.Analyzers.csproj]: are you sure this is supported?

Comment (gradientvera) [reply Robust.Analyzers/Robust.Analyzers.csproj]: works on my machine, but I can revert it easy

Comment (gradientvera) [reply Robust.Analyzers/Robust.Analyzers.csproj]: Reverted, just in case!

File: Robust.Analyzers/FriendAnalyzer.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Analyzers/FriendAnalyzer.cs]: Mmm, technically this should always be fine, I looked at your analyzer and you don't even check if it's null so I'll just remove the check and get it always

**Review by PaulRitter — APPROVED** [review]

**Pull request merged**